pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java

```diff
@@ -0,0 +1,310 @@
+import org.apache.pinot.common.request.context.predicate.NotEqPredicate;
+import org.apache.pinot.common.request.context.predicate.NotInPredicate;
+import org.apache.pinot.common.request.context.predicate.Predicate;
+import org.apache.pinot.segment.local.segment.creator.impl.inv.json.BaseJsonIndexCreator;
+import org.apache.pinot.segment.local.segment.index.readers.BitmapInvertedIndexReader;
+import org.apache.pinot.segment.local.segment.index.readers.StringDictionary;
+import org.apache.pinot.segment.spi.index.reader.JsonIndexReader;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.spi.exception.BadQueryRequestException;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.sql.parsers.CalciteSqlParser;
+import org.roaringbitmap.IntConsumer;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+import static org.apache.pinot.common.request.context.FilterContext.Type.PREDICATE;
+
+
+/**
+ * Reader for json index.
+ */
+public class ImmutableJsonIndexReader implements JsonIndexReader {
+  // NOTE: Use long type for _numDocs to comply with the RoaringBitmap APIs.
+  private final long _numDocs;
+  private final int _version;
+  private final StringDictionary _dictionary;
+  private final BitmapInvertedIndexReader _invertedIndex;
+  private final PinotDataBuffer _docIdMapping;
+
+  public ImmutableJsonIndexReader(PinotDataBuffer dataBuffer, int numDocs) {
+    _numDocs = numDocs;
+    _version = dataBuffer.getInt(0);
+    Preconditions.checkState(_version == BaseJsonIndexCreator.VERSION_1 || _version == BaseJsonIndexCreator.VERSION_2,
+        "Unsupported json index version: %s", _version);
+
+    int maxValueLength = dataBuffer.getInt(4);
+    long dictionaryLength = dataBuffer.getLong(8);
+    long invertedIndexLength = dataBuffer.getLong(16);
+    long docIdMappingLength = dataBuffer.getLong(24);
+
+    long dictionaryStartOffset = BaseJsonIndexCreator.HEADER_LENGTH;
+    long dictionaryEndOffset = dictionaryStartOffset + dictionaryLength;
+    _dictionary =
+        new StringDictionary(dataBuffer.view(dictionaryStartOffset, dictionaryEndOffset, ByteOrder.BIG_ENDIAN), 0,
+            maxValueLength, (byte) 0);
+    long invertedIndexEndOffset = dictionaryEndOffset + invertedIndexLength;
+    _invertedIndex = new BitmapInvertedIndexReader(
+        dataBuffer.view(dictionaryEndOffset, invertedIndexEndOffset, ByteOrder.BIG_ENDIAN), _dictionary.length());
+    long docIdMappingEndOffset = invertedIndexEndOffset + docIdMappingLength;
+    _docIdMapping = dataBuffer.view(invertedIndexEndOffset, docIdMappingEndOffset, ByteOrder.LITTLE_ENDIAN);
+  }
+
+  @Override
+  public MutableRoaringBitmap getMatchingDocIds(String filterString) {
+    FilterContext filter;
+    try {
+      filter = RequestContextUtils.getFilter(CalciteSqlParser.compileToExpression(filterString));
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)